src/ray/gcs/redis_context-inl.h

```diff
@@ -0,0 +1,145 @@
+                   const RedisConnectFunctionType &connect_function) {
+  RAY_LOG(INFO) << "Attempting to connect to address " << address << ":" << port << ".";
+  int connection_attempts = 0;
+  auto resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+  auto status = resp.first;
+  while (!status.ok()) {
+    if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
+      RAY_LOG(FATAL) << RayConfig::instance().redis_db_connect_retries() << " attempts "
+                     << "to connect have all failed. Please check whether the"
+                     << " redis storage is alive or not. The last error message was: "
+                     << status.ToString();
+      break;
+    }
+    RAY_LOG_EVERY_MS(ERROR, 1000)
+        << "Failed to connect to Redis due to: " << status.ToString()
+        << ". Will retry in "
+        << RayConfig::instance().redis_db_connect_wait_milliseconds() << "ms.";
+
+    // Sleep for a little.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        RayConfig::instance().redis_db_connect_wait_milliseconds()));
+    resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+    status = resp.first;
+    connection_attempts += 1;
+  }
+  return resp;
+}
+
+template <typename ConnectType = redisAsyncContext>
+void RedisRequestContext::RedisResponseFn(struct redisAsyncContext *async_context,
+                                          void *raw_reply,
+                                          void *privdata) {
+  auto *request_cxt = static_cast<RedisRequestContext *>(privdata);
+  auto redis_reply = reinterpret_cast<redisReply *>(raw_reply);
+  // Error happened.
+  if (redis_reply == nullptr || redis_reply->type == REDIS_REPLY_ERROR) {
+    auto error_msg = redis_reply ? redis_reply->str : async_context->errstr;
+    RAY_LOG(ERROR) << "Redis request [" << absl::StrJoin(request_cxt->redis_cmds_, " ")
+                   << "]"
+                   << " failed due to error " << error_msg << ". "
+                   << request_cxt->pending_retries_ << " retries left.";
+    if (RayConfig::instance().enable_moved_redirect()) {
+      if (auto maybe_ip_port =
+              ParseIffMovedError(std::string(redis_reply->str, redis_reply->len));
+          maybe_ip_port.has_value()) {
+        const auto [ip, port] = maybe_ip_port.value();
+        auto resp = ConnectWithRetries<ConnectType>(ip.c_str(), port, redisAsyncConnect);
+        if (auto st = resp.first; !st.ok()) {
+          // We will ultimately return a MOVED error if we fail to reconnect.
+          RAY_LOG(ERROR) << "Failed to connect to the new leader " << ip << ":" << port;
+        } else {
+          request_cxt->redis_context_.reset(
+              new RedisAsyncContext(std::move(resp.second)));
+          // TODO(vitsai): do we need to Attach
+        }
+      }
+    }
+    auto delay = request_cxt->exp_back_off_.Current();
+    request_cxt->exp_back_off_.Next();
```

Comment:
NIT, but I think for this error, it's just a redirection, and we should just redirect it and no need to delay anything.

Reply:
done!